test/test_torch.py

```diff
@@ -76,6 +76,26 @@ def filter_supported_types(self, types):
            types = [t for t in types if t in ccl_supported_types]
         return types
 
+    def test_horovod_reinit(self):
+        """Test that Horovod can init -> shutdown -> init successfully."""
+        mpi_rank, _ = mpi_env_rank_and_size()
+        gloo_rank = int(os.getenv('HOROVOD_RANK', -1))
+
+        is_mpi = gloo_rank == -1
+        if is_mpi:
+            # Only applies for Gloo
```

Comment:
Good catch on this one and glad you added a test for reinitialization! Does this issue only impact gloo?

Reply:
In the case of MPI, it seems that shutdown is permanent:

```
*** The MPI_Query_thread() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort
```

The reinit support is relevant for Gloo because we need to do it for fault tolerance.

Reply:
Yep, you are absolutely right. Shutdown invokes `MPI_Finalize` which is permanent. Looks good!